bindings/rust/s2n-tls/src/callbacks/pkey.rs

```diff
@@ -0,0 +1,371 @@
+    pub fn kind(&self) -> Result<&OperationType, Error> {
+        Ok(&self.kind)
+    }
+
+    /// The size of the slice returned by [`input()`]
+    pub fn input_size(&self) -> Result<u32, Error> {
+        let mut size = 0;
+        unsafe { s2n_async_pkey_op_get_input_size(self.raw.as_ptr(), &mut size) }.into_result()?;
+        Ok(size)
+    }
+
+    /// Provides the input for the operation.
+    ///
+    /// If this is an [`OperationType::Sign`] operation, then this input has
+    /// already been hashed and is the resultant digest.
+    pub fn input(&self, buf: &mut [u8]) -> Result<(), Error> {
+        let buf_len: u32 = buf.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        let buf_ptr = buf.as_ptr() as *mut u8;
+        unsafe { s2n_async_pkey_op_get_input(self.raw.as_ptr(), buf_ptr, buf_len) }
+            .into_result()?;
+        Ok(())
+    }
+
+    /// Sets the output of the operation
+    pub fn set_output(self, conn: &mut Connection, buf: &[u8]) -> Result<(), Error> {
+        let buf_len: u32 = buf.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        let buf_ptr = buf.as_ptr() as *const u8;
+        unsafe {
+            s2n_async_pkey_op_set_output(self.raw.as_ptr(), buf_ptr, buf_len).into_result()?;
+            s2n_async_pkey_op_apply(self.raw.as_ptr(), conn.as_ptr()).into_result()?;
+        }
+        Ok(())
+    }
+}
+
+impl Drop for PrivateKeyOperation {
+    fn drop(&mut self) {
+        unsafe {
+            let _ = s2n_async_pkey_op_free(self.raw.as_ptr());
+        }
+    }
+}
+
+pub trait PrivateKeyCallback {
+    fn handle_operation(
+        &self,
+        connection: &mut Connection,
+        operation: PrivateKeyOperation,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, Error>;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        config, connection, error, security, testing,
+        testing::{s2n_tls::*, *},
+    };
+    use core::task::{Poll, Waker};
+    use futures_test::task::new_count_waker;
+    use openssl::{ec::EcKey, ecdsa::EcdsaSig};
+
+    const MAX_POLLS: usize = 100;
+    type Error = Box<dyn std::error::Error>;
+
+    const KEY: &[u8] = include_bytes!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/../../../tests/pems/ecdsa_p384_pkcs1_key.pem"
+    ));
+    const CERT: &[u8] = include_bytes!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/../../../tests/pems/ecdsa_p384_pkcs1_cert.pem"
+    ));
+
+    fn new_pair<T>(
+        callback: T,
+        waker: Waker,
+    ) -> Result<Pair<s2n_tls::Harness, s2n_tls::Harness>, Error>
+    where
+        T: 'static + PrivateKeyCallback,
+    {
+        let config = {
+            let mut config = config::Builder::new();
+            config.set_security_policy(&security::DEFAULT_TLS13)?;
+            config.load_pem(CERT, KEY)?;
+            config.set_private_key_callback(callback)?;
+            unsafe { config.disable_x509_verification() }?;
```

Comment:
Can you add a note as to why this being disabled, especially since we're testing that signatures are working?

Reply:
Added one. Does that clarify things?